### ISSUE TITLE
Inventory Plugin can't cope with IPv6-only VMs

### ISSUE THREAD

**Issue by obsti8383** [ansible-collections/hetzner.hcloud#173]
##### SUMMARY

ansible-inventory and ansible-playbook (when refreshing the inventory) display the following warning as soon as one ipv6-only server is deployed:
 
```
[WARNING]:  * Failed to parse /home/xxx/ansible/hcloud.yml with auto plugin: 'NoneType' object has no attribute 'ip'
[WARNING]:  * Failed to parse /home/xxx/ansible/hcloud.yml with yaml plugin: Plugin configuration YAML file, not YAML inventory
[WARNING]:  * Failed to parse /home/xxx/ansible/hcloud.yml with ini plugin: Invalid host pattern 'plugin:' supplied, ending in ':' is not allowed, this character is reserved to
provide a port.
[WARNING]:  * Failed to parse /home/xxx/ansible/hcloud.yml with ansible_collections.hetzner.hcloud.plugins.inventory.hcloud plugin: 'NoneType' object has no attribute 'ip'
[WARNING]: Unable to parse /home/xxx/ansible/hcloud.yml as an inventory source
```

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
Inventory Plugin 1.9.0

##### ANSIBLE VERSION

```
ansible [core 2.14.0]
  config file = /home/xxx/ansible/ansible.cfg
  configured module search path = ['/home/xxx/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/xxx/.local/lib/python3.10/site-packages/ansible
  ansible collection location = /home/xxx/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/xxx/.local/bin/ansible
  python version = 3.10.2 (main, Jan 24 2022, 21:27:28) [GCC 7.5.0] (/usr/local/bin/python3.10)
  jinja version = 3.1.2
  libyaml = True
```

##### COLLECTION VERSION
```
ansible-galaxy collection list hetzner.hcloud

# /home/xxx/.ansible/collections/ansible_collections
Collection     Version
-------------- -------
hetzner.hcloud 1.9.0 
```

##### CONFIGURATION

```
CONFIG_FILE() = /home/xxx/ansible/ansible.cfg
DEFAULT_HOST_LIST(/home/xxx/ansible/ansible.cfg) = ['/home/xxx/ansible/inventory.yaml', '/home/xxx/ansible/hcloud.yml']
INTERPRETER_PYTHON(/home/xxx/ansible/ansible.cfg) = auto_silent
INVENTORY_ENABLED(/home/xxx/ansible/ansible.cfg) = ['host_list', 'script', 'auto', 'yaml', 'ini', 'toml', 'hcloud']
```

##### OS / ENVIRONMENT
Linux, OpenSuse

##### STEPS TO REPRODUCE
- create IPv6-only VM in Hetzner Cloud
- Execute `ansible-inventory -i hcloud.yml --list`
- Warning output: see above

##### EXPECTED RESULTS

No warnings are shown.

##### ACTUAL RESULTS

The functionality required in my ansible-playbook is working fine (WARNING is not relevant here). However I can't view the inventory with galaxy-inventory at all.

**Comment by apricote** [comment]
Hi @obsti8383

one of the warnings you posted is relevant and the cause for this issue:

```
[WARNING]:  * Failed to parse /home/xxx/ansible/hcloud.yml with auto plugin: 'NoneType' object has no attribute 'ip'
```

this happens because the inventory plugin tries to set the `ansible_host` of the machine to it's (missing) ipv4 address:

https://github.com/ansible-collections/hetzner.hcloud/blob/a3ab379c5b48a6496f588945466a58996eee4cd5/plugins/inventory/hcloud.py#L223-L224

You can use another `connect_with` setting to avoid running into this branch. The default is `public_ipv4`, but depending on your exact use case `hostname` or `private_ipv4` might work instead. You can set this in the `hcloud.yml` file you already have.

---

Furthermore I think we should add support for connecting through IPv6 by adding a new `connect_with` setting called `public_ipv6`.

**Comment by morremeyer** [comment]
I just ran into this after debugging for quite a few minutes. 

Is there a way the plugin can automatically fall back to IPv6 when no IPv4 address is present? 

Or even better in a world that needs to move to IPv6, use the IPv6 address as `ansible_host` by default and falling back to IPv4 if there is no IPv6 address - i.e. changing  the default of `connect_with = "public_ipv4"` to `"public_ipv6"` with a built-in IPv4 fallback.

I can look into opening a PR for this change, but it's been a bit since I wrote code in Python.
